### PR TITLE
Replacing CudaAsyncBuffer with TArray to improve perf

### PR DESCRIPTION
**Pull request by weixingzhang**
**Description**: Describe your changes.
Using TArray instead of CudaAsyncBuffer to passing tensor meta data(mostly, tensor strides/pitches) to CUDA kernels to improve performance

**Motivation and Context**
 Basically, with CudaAsyncBuffer, the sequence at host side is like below. 
1. fill meta data to CPU Pinned memory
2. call cudaMemcpyAsync to upload these meta data from CPU Pinned memory to GPU memory.
3. launch CUDA kernel.

On device(GPU) side, the copy in #2 won't be executed until the moment when kernel in #3 is about to run on GPU. Usually, host side runs far ahead of device side. So, it is better to do data uploading early(during kernel launch time) on host side rather than until kernel execution to save time on GPU. To achieve it, we introduced TArray by which the data is passed to CUDA kernels by pass-by-value. On BERT-L training, we saw about 3-5% perf improvement.

### CODE REVIEW

**Comment by KeDengMS** [comment]
" On BERT-L training, we saw about 3-5% perf improvement." I assume it's for non-fused model right? The change does not seem to touch contrib ops

File: onnxruntime/core/providers/cuda/object_detection/non_max_suppression.cc [hunk 1]
```diff
@@ -94,40 +94,35 @@ Status NonMaxSuppression::ComputeInternal(OpKernelContext* ctx) const {
     Tensor* output = ctx->Output(0, {static_cast<int64_t>(total_num_saved_outputs), last_dim});
     ORT_ENFORCE(output != nullptr);
     int64_t* dst = output->MutableData<int64_t>();
-    size_t count = all_selected_indices.size();
+    int32_t count = static_cast<int32_t>(all_selected_indices.size());
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cuda/object_detection/non_max_suppression.cc]: >all_selected_indices [](start = 41, length = 20)

not sure if the default array size of 8 would be enough. This value depends on selected bounding box, and may pass test but crash in real scene.

File: onnxruntime/core/providers/cuda/rnn/cudnn_rnn_base.cc [hunk 2]
```diff
@@ -348,14 +348,13 @@ Status CudnnRnnBase<T>::ComputeInternal(OpKernelContext* ctx) const {
   }
 
   if ((CUDNN_RNN_RELU == rnn_mode_ || CUDNN_RNN_TANH == rnn_mode_) && sequence_lens_data != nullptr && y_h_data != nullptr && y_data != nullptr) {
-    CudaAsyncBuffer<int32_t> sequence_lens_buffer(this, batch_size);
-    memcpy(sequence_lens_buffer.CpuPtr(), sequence_lens_data, batch_size * sizeof(int32_t));
-    ORT_RETURN_IF_ERROR(sequence_lens_buffer.CopyToGpu());
+    TArray<int32_t> sequence_lens_buffer(batch_size);
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cuda/rnn/cudnn_rnn_base.cc]: >batch_size [](start = 41, length = 10)

not safe for batch_size > 8...I doubt if we have enough test case to catch this though.

**Comment by KeDengMS** [comment]
In general, I have concerns of regression in this change not caught by existing tests. I'd suggest locally change the default TArray size to 4, which should be enough to cover most tensor shape related ops (most ONNX ops assume <= 4D tensors). If any op tests failed, take a close look to decide whether its CudaAsyncBuffer can be replaced by fixed size TArray; Then we can change default TArray size to 8 again.

File: onnxruntime/core/providers/cuda/tensor/concat.cc [hunk 3]
```diff
@@ -59,24 +58,20 @@ Status Concat::ComputeInternal(OpKernelContext* ctx) const {
...
   }
 
-  CudaAsyncBuffer<int64_t> concat_sizes_gpu(this, concat_sizes);
-  CudaAsyncBuffer<int64_t> axis_dimension_input_output_mapping_gpu(this, axis_dimension_input_output_mapping);
-  CudaAsyncBuffer<int64_t> concat_sizes_range_gpu(this, concat_sizes_range);
-  concat_sizes_gpu.CopyToGpu();
-  axis_dimension_input_output_mapping_gpu.CopyToGpu();
-  concat_sizes_range_gpu.CopyToGpu();
-  input_ptr.CopyToGpu();
+  TArray<int64_t> concat_sizes_gpu(concat_sizes);
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cuda/tensor/concat.cc]: >concat_sizes [](start = 35, length = 12)

concat takes variadic inputs, so the default TArray size may not be enough for some models.

File: onnxruntime/core/providers/cuda/tensor/split.cc [hunk 4]
```diff
@@ -47,8 +47,7 @@ Status Split::ComputeInternal(OpKernelContext* ctx) const {
   auto& input_dims = input_shape.GetDims();
   std::vector<int64_t> output_dimensions{input_dims};
 
-  CudaAsyncBuffer<void*> output_ptr(this, num_outputs);
-  gsl::span<void*> output_ptr_span = output_ptr.CpuSpan();
+  TArray<void*> output_ptr(num_outputs);
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cuda/tensor/split.cc]: >num_outputs [](start = 27, length = 11)

Split generates variadic outputs, so fixed size TArray may not be enough.

**Comment by weixingzhang** [comment]
> " On BERT-L training, we saw about 3-5% perf improvement." I assume it's for non-fused model right? The change does not seem to touch contrib ops

Yes. In BERT-L, we only fused Layernorm, GELU, Transpose+Matmul.

**Comment by weixingzhang** [comment]
> In general, I have concerns of regression in this change not caught by existing tests. I had added comments on some ops that fixed sized TArray may overflow.
Agree, I just updated PR to keep CudaAsyncBuffer for non_max_suppression, cudnn_rnn_base, concat, split. Thanks!

**Comment by weixingzhang** [comment]
> In general, I have concerns of regression in this change not caught by existing tests. I had added comments on some ops that fixed sized TArray may overflow.

Agree, I just updated PR to keep CudaAsyncBuffer for non_max_suppression, cudnn_rnn_base, concat, split. Thanks!

**Pull request merged**